sematic/scheduling/kubernetes.py

```diff
@@ -132,6 +189,231 @@ def _unique_job_id_suffix() -> str:
     return uuid.uuid4().hex[:6]
 
 
+def _is_none_or_empty(list_: Optional[List]) -> bool:
+    """
+    Returns True if a list is None or empty, False otherwise.
+    """
+    return list_ is None or len(list_) == 0
+
+
+def _v1_pod_precedence_key(pod: V1Pod) -> int:
+    """
+    To be used as a sorting key when determining the precedence of `V1Pod`s.
+
+    Uses the phase of the pod because the start_time might be None in some phases.
+    """
+    if pod.status is None:
+        return -1
+
+    return (
+        POD_PHASE_PRECEDENCE.index(pod.status.phase)
+        if pod.status.phase in POD_PHASE_PRECEDENCE
+        else -1
+    )
+
+
+def _v1_pod_condition_precedence_key(condition: V1PodCondition) -> Any:
+    """
+    To be used as a sorting key when determining the precedence of `V1PodCondition`s.
+    """
+
+    def condition_cmp(c1: V1PodCondition, c2: V1PodCondition) -> int:
+        c1_phase_index = (
+            POD_CONDITION_PRECEDENCE.index(c1.type)
+            if c1.type in POD_CONDITION_PRECEDENCE
+            else -1
+        )
+
+        c2_phase_index = (
+            POD_CONDITION_PRECEDENCE.index(c2.type)
+            if c2.type in POD_CONDITION_PRECEDENCE
+            else -1
+        )
+
+        return c2_phase_index - c1_phase_index
+
+    return functools.cmp_to_key(condition_cmp)(condition)
+
+
+def _make_final_message(
+    condition: str, reason: Optional[str] = None, message: Optional[str] = None
+) -> str:
+    """
+    Returns a human-readable combination of the three messages.
+    """
+    if not reason and not message:
+        return condition
+    if reason and message:
+        return f"{condition}: {reason}; {message}"
+    return f"{condition}: {reason or message}"
+
+
+def _has_container_failure(container_status: Optional[V1ContainerStatus]) -> bool:
+    """
+    Returns whether the `V1ContainerStatus` object indicates any failure or abnormality.
+    """
+    if container_status is None or container_status.state is None:
+        return True
+    return container_status.state.terminated is not None
+
+
+def _get_standardized_container_state(
+    container_status: Optional[V1ContainerStatus],
+) -> Tuple[str, Optional[str], Optional[str]]:
+    """
+    Tentatively parses out condition, reason, and message strings from the
+    `V1ContainerStatus` object.
+    """
+    if container_status is None or container_status.state is None:
+        return "Container state is unknown!", None, None
+    state = container_status.state
+    if state.waiting is not None:
+        return "Container is waiting", state.waiting.reason, state.waiting.message
+    if state.running is not None:
+        return "Container is running", None, None
+    if state.terminated is not None:
+        return (
+            "Container is terminated",
+            state.terminated.reason,
+            state.terminated.message,
+        )
+    return "Container state is unreadable!", None, None
```

Comment:
That's personal preference but I like to air up the code by adding new lines before `if` statements, especially if the previous line is a `return`. See example in next comment.

Reply:
Done. I also do this sometimes, but I don't have a rule of thumb for when to do it.